lib/Dialect/ESI/ESITranslations.cpp

```diff
@@ -0,0 +1,225 @@
+  return false;
+}
+
+LogicalResult ExportCosimSchema::emitSchemaFor(IntegerType type) {
+  SmallString<16> typeStr;
+  if (type.isSigned())
+    typeStr = "Int";
+  else
+    typeStr = "UInt";
+
+  // Round up.
+  auto w = type.getWidth();
+  if (w == 1)
+    typeStr = "Bool";
+  else if (w <= 8)
+    typeStr += "8";
+  else if (w <= 16)
+    typeStr += "16";
+  else if (w <= 32)
+    typeStr += "32";
+  else if (w <= 64)
+    typeStr += "64";
+  else
+    return diag.emit(unknown, DiagnosticSeverity::Error)
+           << "Capnp does not support ints larger than 64-bit";
+
+  // Since capnp requires messages to be structs, emit a wrapper struct.
+  indent() << "struct ";
+  emitName(type) << " ";
+  emitId(getCapnpTypeID(type)) << " {\n";
+  addIndent();
+
+  // Specify the actual type, followed by the capnp field.
+  indent() << "# Actual type is " << type << ".\n";
+  indent() << "i @0 :" << typeStr << ";\n";
+
+  reduceIndent();
+  indent() << "}\n\n";
+  return success();
+}
+
+LogicalResult ExportCosimSchema::visitEndpoint(CosimEndpoint ep) {
+  ChannelPort inputPort = ep.input().getType().dyn_cast<ChannelPort>();
+  if (!inputPort)
+    return ep.emitOpError("Expected ChannelPort type for input. Got ")
+           << inputPort;
+  if (!isTypeSupported(inputPort))
+    return ep.emitOpError("Type '") << inputPort << "' not supported.";
+  types.insert(inputPort);
+
+  ChannelPort outputPort = ep.output().getType().dyn_cast<ChannelPort>();
+  if (!outputPort)
+    ep.emitOpError("Expected ChannelPort type for output. Got ") << outputPort;
+  if (!isTypeSupported(outputPort))
+    return ep.emitOpError("Type '") << outputPort << "' not supported.";
+  types.insert(outputPort);
+
+  os << "# Endpoint #" << ep.endpointID() << " at " << ep.getLoc() << ":\n";
+  os << "#   Input type: ";
+  emitName(inputPort.getInner()) << " ";
+  emitId(ep.getInputTypeID()) << "\n";
+  os << "#   Output type: ";
+  emitName(outputPort.getInner()) << " ";
+  emitId(ep.getOutputTypeID()) << "\n";
+
+  return success();
+}
+
+LogicalResult ExportCosimSchema::emit() {
+  os << "#########################################################\n"
+     << "## ESI generated schema. For use with CosimDpi.capnp\n"
+     << "#######\n";
+
+  // Walk and collect the type data.
+  module.walk([this](CosimEndpoint ep) { visitEndpoint(ep); });
+  os << "#######\n";
+
+  // We need a sorted list to ensure determinism.
+  SmallVector<Type> sortedTypes(types.begin(), types.end());
+  std::sort(sortedTypes.begin(), sortedTypes.end(), [](Type a, Type b) {
+    return getCapnpTypeID(a) > getCapnpTypeID(b);
+  });
+
+  // Compute and emit the capnp file id.
+  llvm::hash_code fileHash =
+      llvm::hash_combine_range(sortedTypes.begin(), sortedTypes.end());
+  emitId(fileHash) << ";\n\n";
+
+  // Iterate through the various types and emit their schemas.
+  for (auto type : sortedTypes) {
+    auto chanPort = type.dyn_cast<ChannelPort>();
+    if (!chanPort) {
```

Comment:
Isn't it guaranteed the types are all ChannelPort by now? If so, maybe the pass can just assert this invariant rather than emitting diagnostics.

Reply:
In general, I like to program defensively, though an assert would probably have been appropriate. I've switched to storing the ChannelPort instead of a Type, avoiding the issue.